### PR TITLE
Add OCSP_PARTIAL_CHAIN to OCSP_basic_verify()

### PR DESCRIPTION
**Pull request by richsalz**
This adds a flag, OCSP_PARTIAL_CHAIN, to the OCSP_basic_verify()
function.  This is equivalent to X509_V_FLAG_PARTIAL_CHAIN, in that
if any certificate in the OCSP response is in the trust store, then
trust it. The new flag is documented for OCSP.

A new function X509_STORE_CTX_get_flags() was added.  That function is
not documented, since doc/man3/X509_VERIFY_PARAM_set_flags.pod calls
this family out as likely to be deprecated.

-----------

Akamai needs this.  I hope it will get reviewed before the BETA as I think that adding a new flag is a considered to be an API change.

### CODE REVIEW

File: crypto/x509/x509_vfy.c
Comment (DDvO) [inline crypto/x509/x509_vfy.c:2481]: This function is not strictly needed - one can simply combine `X509_STORE_CTX_get0_param()` and `X509_VERIFY_PARAM_get_flags()`.

File: crypto/ocsp/ocsp_vfy.c
Comment (DDvO) [inline crypto/ocsp/ocsp_vfy.c:89]: I just realized that there is quite some overlap between `OCSP_basic_verify()` and `OCSP_request_verify()` regarding the `X509_STORE_CTX` initialization and parameterization.
I suggest extracting a static function for doing the things common to both and using the new flag there,
such that also `OCSP_request_verify()` profits from it.

File: crypto/x509/x509_vfy.c
Comment (richsalz) [reply crypto/x509/x509_vfy.c:2481]: I wanted to keep the layers separate and similar so that if, as documented, stuff gets deprecated, it will be done cleanly.

File: crypto/ocsp/ocsp_vfy.c
Comment (richsalz) [reply crypto/ocsp/ocsp_vfy.c:89]: I don't know, the flags supported are different.  I would really rather not to that in this PR.  I would add the three lines to the other function if you think they are useful.

File: doc/man3/OCSP_resp_find_status.pod
Comment (kroeckx) [inline doc/man3/OCSP_resp_find_status.pod:133]: This looks very confusing to me. I expect that it says that intermediate CAs in the trust store are trust anchors.

Comment (richsalz) [reply doc/man3/OCSP_resp_find_status.pod:133]: The whole paragraph is very poorly written and hard to understand; I didn't want to tackle rewriting it for this PR.

Are you confused that I don't describe the semantics of the new flag, but point to another place?  I really don't want to write something twice, and prefer to point to the definitive source.

Comment (kroeckx) [reply doc/man3/OCSP_resp_find_status.pod:133]: It says "intermediaries that appear in the chain are trusted", which reads to me any intermediary is trusted, whether it's in the trust store or not, and so you're just going to check the end certificate, which just isn't what this does. Either add a few more words to make it clear, or don't describe it at all.

Comment (richsalz) [reply doc/man3/OCSP_resp_find_status.pod:133]: AARgh.  Thanks @kroeckx.  Now I see it.  It should say "intermediates found in the trust store are trusted."  Does that make more sense?

Comment (kroeckx) [reply doc/man3/OCSP_resp_find_status.pod:134]: Yes.

Comment (kroeckx) [reply doc/man3/OCSP_resp_find_status.pod:134]: But you've now said it's about intermediates in the response, which doesn't seem to be the case.

File: crypto/x509/x509_vfy.c
Comment (DDvO) [reply crypto/x509/x509_vfy.c:2481]: I meanwhile had a look at the deprecation note you referred to:

> The above functions should be used to manipulate verification parameters
> instead of functions which work in specific structures such as
> X509_STORE_CTX_set_flags() which are likely to be deprecated in a future release.

I read this as a point **not** to use (or even add) structure-specific functions like `X509_STORE_CTX_*_flags()`,
but for using the generic `X509_VERIFY_PARAM_*` functions, which are orthogonal to the the structures using verification flags.
So please not add to those functions that are likely to get deprecated.

File: crypto/ocsp/ocsp_vfy.c
Comment (DDvO) [reply crypto/ocsp/ocsp_vfy.c:89]: > I don't know, the flags supported are different. I would really rather not to that in this PR. 

Yes, since the flags are different a generalization regarding the verification flags would require an extra flags parameter.

> I would add the three lines to the other function if you think they are useful.

I'm against code redundancy, and copying the new lines to the other place would even add to this.
On the other hand, I meanwhile found that extracting (as I suggested above) a static function for doing the things common to both functions would be way beyond the scope of this PR.
So I'm going to set up a separate PR for cleaning up the OCSP verification code.

File: doc/man3/OCSP_resp_find_status.pod
Comment (richsalz) [reply doc/man3/OCSP_resp_find_status.pod:134]: Intermediates are allowed, and we see them in production, hence this patch.

File: crypto/x509/x509_vfy.c
Comment (richsalz) [reply crypto/x509/x509_vfy.c:2481]: > So please not add to those functions that are likely to get deprecated.

It's an internal use.  It follows what the rest of that code does and is the least invasive change that could be made one week before the beta.
```
; g grep VERIFY_PARAM crypto/ocsp/
        exit 1
;
```

File: crypto/ocsp/ocsp_vfy.c
Comment (richsalz) [reply crypto/ocsp/ocsp_vfy.c:89]: > So I'm going to set up a separate PR for cleaning up the OCSP verification code.

I look forward to that.  Hopefully this PR will get merged and approved. Keep in mind the BETA timetable, but if all you are doing is changing internals (and not API's) you have more time. :) Adding new flags/define's have traditionally been considered an API addition.

File: doc/man3/OCSP_resp_find_status.pod
Comment (kroeckx) [reply doc/man3/OCSP_resp_find_status.pod:134]: It doesn't just affect the intermediates in the response, it also affects intermediates passed in certs.

File: crypto/ocsp/ocsp_vfy.c
Comment (DDvO) [reply crypto/ocsp/ocsp_vfy.c:89]: I've just submitted that follow-up PR: #12669.
It does not touch the API, so is not urgent before the beta.

File: crypto/x509/x509_vfy.c
Comment (DDvO) [reply crypto/x509/x509_vfy.c:2481]: No, you wanna add `X509_STORE_CTX_get_flags()` to the API (without documenting it),
and as mentioned such functions are likely to get deprecated and this function is not needed because
one can use `X509_STORE_CTX_get0_param()` followed by `X509_VERIFY_PARAM_get_flags()` instead.

Comment (richsalz) [reply crypto/x509/x509_vfy.c:2481]: okay, fixup coming.  thanks.

File: doc/man3/OCSP_resp_find_status.pod
Comment (richsalz) [reply doc/man3/OCSP_resp_find_status.pod:134]: And looking at the x509_vfy code, it also accepts the signing cert.  I tried to keep implementation details out of the manpage, and reworded it yet again.  If you have specific suggestions, please let me know.  Thanks for the careful reading.

Comment (kroeckx) [reply doc/man3/OCSP_resp_find_status.pod:134]: We're going to check that `signer` is a certificate we trust. The trusted certs are in `st`, and the untrusted in `untrusted`. `untrusted` is the combination of `certs` and `bs->certs`. The effect of settings partial chain is `st` can contain non-root CAs.

So my suggestion is to just write: If B<flags> contains B<OCSP_PARTIAL_CHAIN>, intermediate CA certificates in B<st> are trust-anchors.

Comment (richsalz) [reply doc/man3/OCSP_resp_find_status.pod:134]: Sure.  Just one other question "If B"  markup issue?  Do you mean "st" ?

Comment (kroeckx) [reply doc/man3/OCSP_resp_find_status.pod:134]: I've put some quotes around it.

Comment (richsalz) [reply doc/man3/OCSP_resp_find_status.pod]: THANKS!  fixup pushed.

**Review by DDvO — APPROVED** [review]
LGTOM

File: crypto/ocsp/ocsp_vfy.c
Comment (kaduk) [inline crypto/ocsp/ocsp_vfy.c:92]: I note that `X509_VERIFY_PARAM_set_flags()` does the "OR-ing in" part for us (it could perhaps be called "add flags" not "set flags"), so the preceding `X509_VERIFY_PARAM_get_flags()` is redundant and gives a confusing impression to the reader of the code.

Comment (richsalz) [reply crypto/ocsp/ocsp_vfy.c:92]: This already got approved, so unless @ddvo says he'll re-review I am not going to make the change.  If nothing else, the code shows how badly-named the API is, since it fooled two seasoned developers. :(

Comment (richsalz) [reply crypto/ocsp/ocsp_vfy.c:91]: Since Akamai decided to make the change, I propagated it here.  I rebased and squashed and asked @ddvo to re-review.  It would be nice if this made it into the BETA, but if not it's on your head, @kaduk :)

File: crypto/ocsp/ocsp_vfy.c
Comment (DDvO) [inline crypto/ocsp/ocsp_vfy.c:91]: Certainly good to make use of the 'or' semantics of ` X509_VERIFY_PARAM_set_flags()`.
Yet I'm puzzled where the `f |` part comes wrong, likely a leftover which should not even compile.

**Review by DDvO — CHANGES_REQUESTED** [review]
I was about to re-approve (BTW please not that my approval weil not be sufficient),
but then I saw that weird `f |` in the updated `X509_VERIFY_PARAM_set_flags()` call.

File: crypto/ocsp/ocsp_vfy.c
Comment (richsalz) [reply crypto/ocsp/ocsp_vfy.c]: gaa, didn't save the buffer before doing the commit.  updated commit pushed.

**Comment by DDvO** [comment]
Re-approved.

**Review by DDvO — APPROVED** [review]

**Review by t8m — APPROVED** [review]
LGTM

**Comment by DDvO** [comment]
@richsalz this is ready to merge :)

**Comment by richsalz** [comment]
@ddvo, you're a committer, I'm not :)

**Comment by DDvO** [comment]
Ah yes, I realized about a second after I hid the "Comment" button that you are not an OpenSSL member any more.
So I've meanwhile done the merge.

**Pull request closed without merging**
